<div align="center">
    <img src="images/zb-logo.png" alt="제로베이스" width="100%">
    <br>
    <a href="https://zerobaseschool.notion.site/DS-EDA-Lv-Test-b7d5e895c7cd4dda813c8fe2f2c23f79"><img src="images/openinnotion_logo.png" alt="EDA Lv Test Doc" width="12%"></a>

이 Jupyter Notebook 파일은 <a href="https://zero-base.co.kr/category_data_camp/school_DS">제로베이스</a>에서 제공하는 **EDA Level Test** 문제입니다.<br>
지금까지 공부하신 내용으로 데이터 분석에 대해 좀 더 연습할 수 있는 시간이 되셨으면 좋겠습니다.🙂<br>
**EDA Level Test**에 대한 안내 사항은 상단 **Open in ZeroBase Notion** 로고를 클릭하여 확인해주세요!<br>
반드시 시작 전에 내용을 숙지해주시기를 바랍니다!<br>
</div>

***

<h1 style="background-color:#9F81F7;color:#D8D8D8;font-weight:bold;display:inline-block;padding: 5px">EDA Level Test 06 ⭐️⭐️</h1>

| Week | Data Analysis | Difficulty | This Notebook |
|------|---------------|------------|-------|
| 1    | 서울시 인구 데이터 | ⭐️ ||
| 2    | 화장품 성분 데이터 | ⭐️⭐️⭐️ ||
| 3    | 올림픽 데이터 | ⭐️⭐️ ||
| 4    | 국가별 인터넷 사용률 데이터 |⭐️⭐️ ||
| 5    | 전국 박물관/미술관 정보 표준 데이터 | ⭐️⭐️⭐️ ||
| 6    | 서울시 흡연율 통계 데이터 | ⭐️⭐️ |👈|
| 7    | 웹크롤링 | ⭐️ ||
| 8    | 서울시 지하철 호선별 역별 유/무임 승하차 인원 정보 데이터 | ⭐️⭐️ ||
| 9    | 서울시 교통사고 현황 (사고유형별) 통계 데이터 | ⭐️⭐️ ||
| 10   | 서울시 교통사고 현황 (사고유형별) 통계 + 서울시 차량통행속도 (구별/월별) 통계 + 서울시 인구밀도 (구별) 통계 | ⭐️⭐️⭐️ ||

## 문제 소개 및 데이터 준비 단계

<div style="border: 1px solid black; padding: 10px;">

### 문제 소개
- 총 3단계 데이터 분석 상황 제시
<br></br>
    - 1단계: DataFrame 전처리 01 - 기초
    - 2단계: DataFrame 전처리 02 - 심화
    - 3단계: 원하는 정보 얻기
<br></br>
    - 총점 100점

### Data 원본 출처

### Target Data(txt): 서울시 흡연율 통계
- Source: [서울 열린데이터 광장](https://data.seoul.go.kr/dataList/10668/S/2/datasetView.do;jsessionid=C18A51D1DF6D7CE1CD05D8C842510242.new_portal-svr-11)
- Reference: [PinkWink](https://pinkwink.kr/1288?category=580892)
- DownLoad: 상기 Source(서울 열린데이터 광장)에서 2009년 - 2018년 설정하여 Download
    - 문제 출제 이후 Data 변경될 가능성이 있으므로 문제 풀이시 하기 File Path의 Data 사용
- File Path: ```'./datas/report.txt'```

### 참고사항
- 서울 열린데이터 광장에서 해당 데이터는 다양한 Format(xls, txt, hwp)으로 제공합니다.
- 데이터는 탭으로 구분하여 TXT로 제공합니다
- 문제에 hint가 있을 경우, 해당 hint를 이용하지 않으셔도 무방합니다.

</div>

In [1]:
import pandas as pd
# 채점을 위한 코드입니다. 반드시 실행해주세요.
from grading import *

df_target = pd.read_csv('./datas/report.txt', sep='\t')
df_target.head()

기간     구분    구분.1    전체    남자   여자
0  2009    서울시     서울시  24.3  45.5    4
1  2009  생애주기별  19~29세  24.9  43.8    6
2  2009  생애주기별  30~44세  29.9  55.1  3.8
3  2009  생애주기별  45~64세  22.6  43.2  3.3
4  2009  생애주기별  65세 이상  11.5  22.8  2.9

## 문제 시작! 🏃🏻

## 1단계: DataFrame 전처리 01 - 기초
***

### 문제 1-1) 기초 전처리 01 (5점) 
- 위의 DataFrame(df_target)을 아래의 조건에 맞게 변경하세요.
<br></br>
    - 조건1: '구분' Column(열)명을 '구분1'로, '구분.1' Column(열)명을 '구분2'로 변경하세요.
<br></br>
    - 조건2: 조건1에 따라 Column(열)명을 변경한 후 '구분1'의 '서울시'를 '시'로, '**구'(예: 양천구, 영등포구)는 '구'로 변경하세요.
<br></br>
    - 조건3: Index, Column 순서 또는 Data 순서(정렬)는 변경하지 마세요.
<br></br>
    - 조건4: 'df_target' 변수에 결과 DataFrame을 할당하세요.
<br></br>
<br></br>
- 완료 후 결과 DataFrame 변수(df_target)를 check_01_01 함수에 입력하여 채점하세요.

In [2]:
# column명 변경 후 check

df_target.rename(columns={'구분': '구분1', '구분.1':'구분2'}, inplace=True)
df_target.head()

기간    구분1     구분2    전체    남자   여자
0  2009    서울시     서울시  24.3  45.5    4
1  2009  생애주기별  19~29세  24.9  43.8    6
2  2009  생애주기별  30~44세  29.9  55.1  3.8
3  2009  생애주기별  45~64세  22.6  43.2  3.3
4  2009  생애주기별  65세 이상  11.5  22.8  2.9

In [3]:
# 구분1 컬럼의 서울시 >> 시로 변경

df_target['구분1'] = df_target['구분1'].replace('서울시', '시')
df_target.head()

기간    구분1     구분2    전체    남자   여자
0  2009      시     서울시  24.3  45.5    4
1  2009  생애주기별  19~29세  24.9  43.8    6
2  2009  생애주기별  30~44세  29.9  55.1  3.8
3  2009  생애주기별  45~64세  22.6  43.2  3.3
4  2009  생애주기별  65세 이상  11.5  22.8  2.9

In [4]:
# .str.contains()로 '구' 찾아서 모두 '구'로 바꿔주기

df_target.loc[df_target['구분1'].str.contains('구'), '구분1'] = '구'

In [5]:
df_target.구분1.unique()

array(['시', '생애주기별', '교육수준별Ⅰ(30~64세)', '교육수준별Ⅱ(65세 이상)', '직업별(30~64세)',
       '구'], dtype=object)

In [6]:
# 1-1

check_01_01(df_target)
df_target.head()

정답입니다! 5점 누적 되었습니다!
현재 누적 점수: 5 / 100


기간    구분1     구분2    전체    남자   여자
0  2009      시     서울시  24.3  45.5    4
1  2009  생애주기별  19~29세  24.9  43.8    6
2  2009  생애주기별  30~44세  29.9  55.1  3.8
3  2009  생애주기별  45~64세  22.6  43.2  3.3
4  2009  생애주기별  65세 이상  11.5  22.8  2.9

### 문제 1-2) 기초 전처리 02 (10점) 
- 1-1의 DataFrame(df_target)을 아래의 조건에 맞게 변경하세요.
<br></br>
    - 조건1: '여자' Column(열)명의 Data 중 float Type으로 변경할 수 있는 경우 float(numpy.float64)으로 Type을 변경하고, 불가능한 경우 해당 Data의 값을 Null값(None)으로 변경하세요.
<br></br>
    - 조건2: Index, Column 순서 또는 Data 순서(정렬)는 변경하지 마세요.
<br></br>
    - 조건3: 'df_target' 변수에 결과 DataFrame을 할당하세요.
<br></br>
<br></br>
    - hint1: python 예외처리(try-except)를 활용해보세요.
<br></br>
    - hint2: 현재 df_target의 '여자' Column(열) data들의 Type은 str입니다.
<br></br>
<br></br>
- 완료 후 결과 DataFrame 변수(df_target)를 check_01_02 함수에 입력하여 채점하세요.

In [7]:
# 테스트를 위해 카피본 생성

df_test = df_target.copy()

In [8]:
# 데이터 타입 확인

type(df_test.여자[0])

str

In [9]:
# 먼저 float으로 바꿀 수 있는 숫자들은 float으로 바꾸고 아닌 것은 None으로 변경

for idx, rows in df_test.iterrows():
    try:
        df_test.여자[idx] = float(df_test.여자[idx])
    except:
        df_test.여자[idx] = None

C:\Users\User\AppData\Local\Temp\ipykernel_18880\1423741054.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_test.여자[idx] = float(df_test.여자[idx])
C:\Users\User\AppData\Local\Temp\ipykernel_18880\1423741054.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_test.여자[idx] = None


In [10]:
# 이후 float64로 열 전체 데이터 타입 변경 (None으로 바뀔 것은 바뀌었기 때문에 error 안남)

df_test.여자 = df_test.여자.astype('float64')

In [11]:
# 카피본을 다시 베껴와서 target에 할당

df_target = df_test.copy()

In [12]:
# 1-2

check_01_02(df_target)
df_target.head()

정답입니다! 10점 누적 되었습니다!
현재 누적 점수: 15 / 100


기간    구분1     구분2    전체    남자   여자
0  2009      시     서울시  24.3  45.5  4.0
1  2009  생애주기별  19~29세  24.9  43.8  6.0
2  2009  생애주기별  30~44세  29.9  55.1  3.8
3  2009  생애주기별  45~64세  22.6  43.2  3.3
4  2009  생애주기별  65세 이상  11.5  22.8  2.9

### 문제 1-3) 기초 전처리 03 (5점) 
- 1-2의 DataFrame(df_target)을 아래의 조건에 맞게 변경하세요.
<br></br>
    - df_target의 '구분' Column(열) Data 중 '교육수준별Ⅱ(65세 이상)'의 Data들은 Null값이 너무 많아 분석 대상에서 제외하려고 합니다.
<br></br>
<br></br>
    - 조건1: '구분' Column(열)이 '교육수준별Ⅱ(65세 이상)'인 Row(행)을 삭제하세요.
<br></br>
    - 조건2: Index, Column 순서 또는 Data 순서(정렬)는 변경하지 마세요.
<br></br>
    - 조건3: 'df_target' 변수에 결과 DataFrame을 할당하세요.
<br></br>
<br></br>
- 완료 후 결과 DataFrame 변수(df_target)를 check_01_02 함수에 입력하여 채점하세요.

In [13]:
# 데이터프레임 재 체크

df_target.head()

기간    구분1     구분2    전체    남자   여자
0  2009      시     서울시  24.3  45.5  4.0
1  2009  생애주기별  19~29세  24.9  43.8  6.0
2  2009  생애주기별  30~44세  29.9  55.1  3.8
3  2009  생애주기별  45~64세  22.6  43.2  3.3
4  2009  생애주기별  65세 이상  11.5  22.8  2.9

In [14]:
df_target.구분1.unique()

array(['시', '생애주기별', '교육수준별Ⅰ(30~64세)', '교육수준별Ⅱ(65세 이상)', '직업별(30~64세)',
       '구'], dtype=object)

In [15]:
df_target[df_target['구분1'] == '교육수준별Ⅱ(65세 이상)']

기간             구분1   구분2    전체    남자   여자
8    2009  교육수준별Ⅱ(65세 이상)  초졸이하   8.9  25.0  NaN
9    2009  교육수준별Ⅱ(65세 이상)    중졸  15.1  28.0  NaN
10   2009  교육수준별Ⅱ(65세 이상)  고졸이상  14.1  20.1  NaN
47   2010  교육수준별Ⅱ(65세 이상)  초졸이하   9.6  25.5  NaN
48   2010  교육수준별Ⅱ(65세 이상)    중졸  13.1  23.2  NaN
49   2010  교육수준별Ⅱ(65세 이상)  고졸이상  13.4  19.6  NaN
86   2011  교육수준별Ⅱ(65세 이상)  초졸이하   9.1  24.2  NaN
87   2011  교육수준별Ⅱ(65세 이상)    중졸  11.2  20.5  NaN
88   2011  교육수준별Ⅱ(65세 이상)  고졸이상  11.9  17.2  NaN
125  2012  교육수준별Ⅱ(65세 이상)  초졸이하   7.7  20.8  NaN
126  2012  교육수준별Ⅱ(65세 이상)    중졸  13.7  25.7  NaN
127  2012  교육수준별Ⅱ(65세 이상)  고졸이상  13.3  20.0  NaN
164  2013  교육수준별Ⅱ(65세 이상)  초졸이하   8.3  23.3  NaN
165  2013  교육수준별Ⅱ(65세 이상)    중졸  10.6  19.3  NaN
166  2013  교육수준별Ⅱ(65세 이상)  고졸이상  12.9  18.0  NaN
203  2014  교육수준별Ⅱ(65세 이상)  초졸이하   8.1  22.2  NaN
204  2014  교육수준별Ⅱ(65세 이상)    중졸  11.6  22.9  NaN
205  2014  교육수준별Ⅱ(65세 이상)  고졸이상  12.0  18.0  NaN
242  2015  교육수준별Ⅱ(65세 이상)  초졸이하   5.5  15.8  NaN
243  2015  교육수준별Ⅱ(65세 이상)    중졸   9.0  16.6  NaN
244  2015  교육수준별Ⅱ(65세 이상)  고졸이상  11.3  17.4  NaN
281  2016  교육수준별Ⅱ(65세 이상)  초졸이하   7.3  22.2  1.9
282  2016  교육수준별Ⅱ(65세 이상)    중졸   9.4  16.8  2.4
283  2016  교육수준별Ⅱ(65세 이상)  고졸이상  11.5  17.8  NaN
320  2017  교육수준별Ⅱ(65세 이상)  초졸이하   6.9  19.1  NaN
321  2017  교육수준별Ⅱ(65세 이상)    중졸   9.6  19.4  NaN
322  2017  교육수준별Ⅱ(65세 이상)  고졸이상   9.3  14.7  NaN
359  2018  교육수준별Ⅱ(65세 이상)  초졸이하   6.5  19.1  1.8
360  2018  교육수준별Ⅱ(65세 이상)    중졸  11.5  21.9  1.6
361  2018  교육수준별Ⅱ(65세 이상)  고졸이상  10.4  16.2  1.4
398  2019  교육수준별Ⅱ(65세 이상)  초졸이하   6.6  19.0  1.7
399  2019  교육수준별Ⅱ(65세 이상)    중졸  10.6  19.2  2.2
400  2019  교육수준별Ⅱ(65세 이상)  고졸이상   9.8  15.7  1.0

In [16]:
# 교육수준별Ⅱ(65세 이상) 을 갖는 행 인덱스를 구하여 drop
# 해당 행 있는지 재 확인

df_target.drop(df_target[df_target['구분1'] == '교육수준별Ⅱ(65세 이상)'].index, axis=0, inplace=True)
df_target[df_target['구분1'] == '교육수준별Ⅱ(65세 이상)']

Empty DataFrame
Columns: [기간, 구분1, 구분2, 전체, 남자, 여자]
Index: []

In [17]:
# 1-3

check_01_03(df_target)
df_target.head()

정답입니다! 5점 누적 되었습니다!
현재 누적 점수: 20 / 100


기간    구분1     구분2    전체    남자   여자
0  2009      시     서울시  24.3  45.5  4.0
1  2009  생애주기별  19~29세  24.9  43.8  6.0
2  2009  생애주기별  30~44세  29.9  55.1  3.8
3  2009  생애주기별  45~64세  22.6  43.2  3.3
4  2009  생애주기별  65세 이상  11.5  22.8  2.9

## 2단계: DataFrame 전처리 02 - 심화
***

### 문제 2-1) 심화 전처리 01 (10점) 
- 1단계의 DataFrame(df_target)을 아래의 조건에 맞게 변경하세요.
<br></br>
    - df_target '여자' Column(열)의 Data에 Null값을 채우고자 합니다. 
<br></br>
    - Null값을 채우는 방법으로는 여러가지가 있는데, 이 Test에서는 '전체'와 '남자', '여자' Column(열)의 Data를 이용하여 '여자비율'을 구하고, '여자비율'의 변화를 확인하여 Null값을 채워보려고 합니다.
<br></br>
    - **df_target.전체 = df_target.여자 * x/100 + df_target.남자 * (1 - x/100)** 에서의 x% 값이 '여자비율'입니다.
<br></br>
    - 예시
<br></br>
    <img src="images/06-2-1-001.png" style="margin:auto" width=400/>
<br></br>
<br></br>
    - 조건1: '여자' Column(열)의 Data가 Null값(None)이 아닌 경우 위의 수식에 따른 '여자비율'을 구하여 '여자비율' Column에 입력하세요(위 예시를 참고).
        - '여자비율' Column(열)의 Data는 소숫점 일곱 번째 자리에서 반올림하여 소숫점 여섯 번째까지 입력해주세요.
        - '여자비율' Column(열) Data의 Type은 float(numpy.float64)이어야 합니다.
<br></br>
    - 조건2: '여자' Column(열)의 Data가 Null값(None)인 경우 '여자비율' Column에 Null값(None)을 입력하세요(위 예시를 참고).
<br></br>
    - 조건3: Index, Column 순서 또는 Data 순서(정렬)는 변경하지 마세요.
<br></br>
    - 조건4: 'df_target' 변수에 결과 DataFrame을 할당하세요.
<br></br>
<br></br>
    - hint1: 방정식을 풀기 위한 Python Library로 Sympy 등이 있습니다. 해당 문제는 sympy를 이용해서 풀수도 있고, 비교적 쉬운 1차 방정식이므로 수식을 직접 전개해서 풀어도 상관 없습니다.
<br></br>
<br></br>
- 완료 후 결과 dataframe 변수(df_target)를 check_02_01 함수에 입력하여 채점하세요.

In [18]:
# !pip install sympy

In [19]:
import sympy as sy
import numpy as np

In [20]:
# sympy test

x = sy.symbols('x')

f = sy.Eq(df_target.여자[0] * x/100 + df_target.남자[0] * (1 - x/100), df_target.전체[0])

result = sy.solve(f)
result

[51.0843373493976]

In [21]:
# sympy 적용하여 각 행에 결과 대입

x = sy.symbols('x')
df_target['여자비율'] = np.nan

for idx, rows in df_target.iterrows():
    try:
        f = sy.Eq(df_target.여자[idx] * x/100 + df_target.남자[idx] * (1 - x/100), df_target.전체[idx])
        df_target['여자비율'][idx] = sy.solve(f)[0]
    except:
        df_target['여자비율'][idx] = None

C:\Users\User\AppData\Local\Temp\ipykernel_18880\941804776.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_target['여자비율'][idx] = sy.solve(f)[0]


In [22]:
# sympy 결과는 float64가 아니기 때문에 데이터 타입 변환 및, 반올림

df_target.여자비율 = df_target.여자비율.astype('float64').round(6)
df_target.head()

기간    구분1     구분2    전체    남자   여자       여자비율
0  2009      시     서울시  24.3  45.5  4.0  51.084337
1  2009  생애주기별  19~29세  24.9  43.8  6.0  50.000000
2  2009  생애주기별  30~44세  29.9  55.1  3.8  49.122807
3  2009  생애주기별  45~64세  22.6  43.2  3.3  51.629073
4  2009  생애주기별  65세 이상  11.5  22.8  2.9  56.783920

In [23]:
# 2-1

check_02_01(df_target)
display(df_target.head())

정답입니다! 10점 누적 되었습니다!
현재 누적 점수: 30 / 100


기간    구분1     구분2    전체    남자   여자       여자비율
0  2009      시     서울시  24.3  45.5  4.0  51.084337
1  2009  생애주기별  19~29세  24.9  43.8  6.0  50.000000
2  2009  생애주기별  30~44세  29.9  55.1  3.8  49.122807
3  2009  생애주기별  45~64세  22.6  43.2  3.3  51.629073
4  2009  생애주기별  65세 이상  11.5  22.8  2.9  56.783920

### 문제 2-2) 심화 전처리 02 (10점) 
- 2-1의 DataFrame(df_target)을 이용하여 아래의 조건에 맞는 정보를 구하여 DataFrame(df_result)으로 나타내세요.
<br></br>
    - df_target의 '여자' Column Data의 Null값(None)을 채우기 위한 근거로 아래와 같은 정보를 구하려고 합니다.
<br></br>
    - 2-1에서 구한 df_target '여자비율' Column Data의 연간 변동 정도를 '구분2' Column(열) Data 중 '여자비율'이 Null값(None)이 있는 '구분2' Data 별로 다양한 방법(회귀계수(coefficient), MinMax, 평균, 표준편차)을 이용하여 확인해보려 합니다.
<br></br>
    - 회귀계수란?
        - 선형 모형에서 x와 y의 관계를 수식으로 y = w * x + b 로 나타낼 수 있습니다.
        - 이 식에서 w(기울기)와 b(절편)를 회귀계수(cofficient)라고 합니다.
        - 독립변수 x가 1 증가할 때마다 종속변수 y는 w만큼 증가합니다.
        - 아래의 조건2에서 우리가 구하고자 하는 것은 회귀계수 중 기울기(w)이며, 편의상 w를 '회귀계수'라 표현하겠습니다.
<br></br>
    - 아래 예시는 아래의 조건에 맞추어 '전체' Column(열) Data의 연간 변동을 '구분2' Column(열) 별로 확인해 본 것 입니다.
<br></br>
        - 예시1
<br></br>
            <img src="images/06-2-2-001.png" style="margin:auto" width=400/>
<br></br>
<br></br>
    - 조건1: 결과 DataFrame의 Column(열)명은 ['구', '회귀계수', 'MinMax', '평균', '표준편차'] 로 만들어주세요.
<br></br>
    - 조건2: '여자' 또는 '여자비율' Column(열) Data가 Null값(None)이 아닌 Row(행)의 '구분2' Column(열) Data를 구하고, 오름차순으로 정렬하세요.
<br></br>
        - 예시2: df_target의 일부입니다. 아래 그림에 대한 조건2의 결과는 [노원구, 동작구, 서초구] 입니다.
<br></br>
        <img src="images/06-2-2-002.png" style="margin:auto" width=400/>
<br></br>
    - 조건3: 조건2에 해당하는 Data 각각 '기간' Column(열) Data의 변화에 따른 '여자비율' Column(열)의 변동을 나타내는 '회귀계수'를 구하세요. 이 결과를 결과 DataFrame의  '회귀계수' Column(열)에 입력하세요.
        - 단순회귀분석(1차, 직선)의 회귀계수 중 기울기를 의미합니다. 위의 '회귀계수란?'을 참고해주세요.
        - 위 그림의 예시2에 따르면 [노원구, 동작구, 서초구] 각각에 대한 기간별 '회귀계수'를 구하면 됩니다.
        - '기간' Column(열)의 Data가 x가 됩니다.
        - '여자비율' Column(열) Data 가 y가 됩니다.
        - 소숫점 일곱번째 자리에서 반올림하여 여섯번째 자리까지만 나타내세요.
        - 당연히 Null값(None)이 있는 Row(행)은 계산시 제외됩니다.
<br></br>
    - 조건4: 조건2에 해당하는 Data 각각 '여자비율' Column(열) Data의 최대값과 최소값의 차이(MinMax)를 구하세요. 이 결과를 결과 DataFrame의  'MinMax' Column(열)에 입력하세요.
        - 위 그림의 예시2에 따르면 [노원구, 동작구, 서초구] 각각에 대한 최대값과 최소값의 차이를 구하면 됩니다.
        - 소숫점 일곱번째 자리에서 반올림하여 여섯번째 자리까지만 나타내세요.
        - 절대값(양수)으로 표현하세요.
        - 당연히 Null값(None)이 있는 Row(행)은 계산시 제외됩니다.
<br></br>
    - 조건5: 조건2에 해당하는 Data 각각 '여자비율' Column(열)의 평균을 구하세요. 이 결과를 결과 DataFrame의 '평균' Column(열)에 입력하세요.
        - 위 그림의 예시2에 따르면 [노원구, 동작구, 서초구] 각각에 대한 평균을 구하면 됩니다.
        - 소숫점 일곱번째 자리에서 반올림하여 여섯번째 자리까지만 나타내세요.
        - 절대값(양수)으로 표현하세요.
        - 당연히 Null값(None)이 있는 Row(행)은 계산시 제외됩니다.
<br></br>
    - 조건5: 조건2에 해당하는 Data 각각 '여자비율' Column(열)의 평균을 구하세요. 이 결과를 결과 DataFrame의 '표준편차' Column(열)에 입력하세요.
        - 위 그림의 예시2에 따르면 [노원구, 동작구, 서초구] 각각에 대한 표준편차을 구하면 됩니다.
        - 소숫점 일곱번째 자리에서 반올림하여 여섯번째 자리까지만 나타내세요.
        - 절대값(양수)으로 표현하세요.
        - 당연히 Null값(None)이 있는 Row(행)은 계산시 제외됩니다.
<br></br>
    - 조건6: 조건2의 Data를 순서대로 조건3~5의 값을 구하고 DataFrame을 만든다면, 결과 DataFrame의 '구' Column(열)을 기준으로 오름차순으로 정렬되어 있을 것 입니다. 만약 정렬되어있지 않다면, '구'를 기준으로 정렬하고 Index를 초기화하세요.
<br></br>
    - 조건7: 'df_result' 변수에 결과 DataFrame을 할당하세요.
<br></br>
<br></br>
    - hint1: 조건3 관련하여 이전에 배웠던 np.polyfit 함수를 응용해보세요.
<br></br>
    - hint2: 조건3 관련하여 정답 중 '강동구'의 '회귀계수'는 0.052052 입니다.
<br></br>
    - hint3: 예시1의 그림과 Index/Column이 동일합니다. 안의 숫자만 다릅니다.
<br></br>
<br></br>
- 완료 후 결과 dataframe 변수(df_result)를 check_02_02 함수에 입력하여 채점하세요.

In [24]:
# 조건1 특정 이름의 컬럼을 가진 데이터프레임 생성
df_result = pd.DataFrame(columns = ['구', '회귀계수', 'MinMax', '평균', '표준편차'])
df_result

Empty DataFrame
Columns: [구, 회귀계수, MinMax, 평균, 표준편차]
Index: []

In [25]:
# 먼저 구 데이터만 추출

df_test = df_target[df_target.구분1 == '구']
df_test

기간 구분1  구분2    전체    남자   여자       여자비율
14   2009   구  종로구  23.0  41.2  5.3  50.696379
15   2009   구   중구  27.0  48.7  5.3  50.000000
16   2009   구  용산구  25.1  46.2  5.4  51.715686
17   2009   구  성동구  25.3  47.3  3.5  50.228311
18   2009   구  광진구  24.3  44.0  5.4  51.036269
..    ...  ..  ...   ...   ...  ...        ...
424  2019   구  관악구  18.0  30.3  5.6  49.797571
425  2019   구  서초구  12.6  25.1  1.5  52.966102
426  2019   구  강남구  14.7  26.6  4.2  53.125000
427  2019   구  송파구  14.9  28.8  NaN        NaN
428  2019   구  강동구  18.5  34.8  2.9  51.097179

[275 rows x 7 columns]

In [26]:
# 추출한 데이터 중 nan 값이 있는 구 확인

df_test[df_test['여자'].isnull()].구분2.unique()

array(['동대문구', '노원구', '동작구', '송파구', '양천구', '서초구', '성동구', '도봉구', '서대문구',
       '구로구', '영등포구', '강동구'], dtype=object)

In [27]:
# 위에서 확인 된 구 이름 리스트화 (정렬도 미리 진행)

gu_list = list(df_test[df_test['여자'].isnull()].구분2.unique())
gu_list.sort()
gu_list

['강동구',
 '구로구',
 '노원구',
 '도봉구',
 '동대문구',
 '동작구',
 '서대문구',
 '서초구',
 '성동구',
 '송파구',
 '양천구',
 '영등포구']

In [28]:
# 1개 구로 테스트

df_tmp = df_test[df_test.구분2 == '강동구']
df_tmp2 = df_tmp[df_tmp.여자.notnull()]
df_tmp2

기간 구분1  구분2    전체    남자   여자       여자비율
38   2009   구  강동구  24.3  43.9  5.1  50.515464
77   2010   구  강동구  26.7  48.4  5.5  50.582751
116  2011   구  강동구  21.9  40.2  3.9  50.413223
155  2012   구  강동구  23.1  42.2  4.3  50.395778
194  2013   구  강동구  23.0  42.3  4.3  50.789474
233  2014   구  강동구  18.5  33.8  3.5  50.495050
272  2015   구  강동구  19.5  36.0  3.3  50.458716
311  2016   구  강동구  18.3  33.9  3.1  50.649351
350  2017   구  강동구  17.4  31.2  4.1  50.922509
428  2019   구  강동구  18.5  34.8  2.9  51.097179

In [29]:
fp1 = np.polyfit(df_tmp2.기간, df_tmp2.여자비율, 1)
fp1[0]

0.05205211038961785

In [30]:
# for문 이용하여 진행

reg = []
minmax = []
avg = []
stdd = []

for i in gu_list:
    df_tmp = df_test[df_test.구분2 == i]
    df_tmp = df_tmp[df_tmp.여자.notnull()]

    fp1 = np.polyfit(df_tmp.기간, df_tmp.여자비율, 1)
    reg.append(fp1[0])
    minmax.append(abs(df_tmp.여자비율.max() - df_tmp.여자비율.min()))
    avg.append(df_tmp.여자비율.mean())
    stdd.append(df_tmp.여자비율.std(ddof=0))
    

In [31]:
# 회귀계수 (기울기) 리스트

reg

[0.05205211038961785,
 0.08509291108892321,
 0.016932133333339185,
 0.024618918273642775,
 0.04268175757576404,
 0.054593320224715046,
 0.059955801652893874,
 0.06607075666666852,
 0.09743204499541211,
 -0.01950991428571394,
 0.024970324074079717,
 0.08297813407821908]

In [32]:
# min max 구하는 법 >> 위 for문에 추가

abs(df_tmp.여자비율.max() - df_tmp.여자비율.min())

0.8896799999999985

In [33]:
# 평균 구하는 법 >> 위 for문에 추가

df_tmp.여자비율.mean()

50.26080722222222

In [34]:
# 표준편차 구하는 법 >> ddof=0 이어야 답과 같은 표준편차 값 나옴 >> 위 for문에 추가

df_tmp = df_test[df_test.구분2 == '도봉구']
df_tmp.여자비율.std(ddof=0)

0.14810022177012452

In [35]:
# 위 for문에서 생성한 리스트들 데이터프레임에 입력

df_result['구'] = gu_list
df_result.회귀계수 = reg
df_result.MinMax = minmax
df_result.평균 = avg
df_result.표준편차 = stdd
df_result = df_result.round(6)
df_result

구      회귀계수    MinMax         평균      표준편차
0    강동구  0.052052  0.701401  50.631950  0.222337
1    구로구  0.085093  1.011929  50.267907  0.293120
2    노원구  0.016932  0.400031  52.033941  0.153785
3    도봉구  0.024619  0.580787  51.265114  0.148100
4   동대문구  0.042682  0.488017  50.060631  0.141627
5    동작구  0.054593  0.531461  51.427075  0.186288
6   서대문구  0.059956  0.935272  51.826214  0.264601
7    서초구  0.066071  0.841456  52.552076  0.250892
8    성동구  0.097432  1.129327  50.518469  0.344289
9    송파구 -0.019510  0.285715  51.585111  0.095467
10   양천구  0.024970  0.378362  51.171848  0.124261
11  영등포구  0.082978  0.889680  50.260807  0.280208

In [36]:
# 2-2

check_02_02(df_result)
df_result

정답입니다! 10점 누적 되었습니다!
현재 누적 점수: 40 / 100


구      회귀계수    MinMax         평균      표준편차
0    강동구  0.052052  0.701401  50.631950  0.222337
1    구로구  0.085093  1.011929  50.267907  0.293120
2    노원구  0.016932  0.400031  52.033941  0.153785
3    도봉구  0.024619  0.580787  51.265114  0.148100
4   동대문구  0.042682  0.488017  50.060631  0.141627
5    동작구  0.054593  0.531461  51.427075  0.186288
6   서대문구  0.059956  0.935272  51.826214  0.264601
7    서초구  0.066071  0.841456  52.552076  0.250892
8    성동구  0.097432  1.129327  50.518469  0.344289
9    송파구 -0.019510  0.285715  51.585111  0.095467
10   양천구  0.024970  0.378362  51.171848  0.124261
11  영등포구  0.082978  0.889680  50.260807  0.280208

### 문제 2-3) 심화 전처리 03 (10점) 
- 2-2의 DataFrame(df_result)을 이용하여 아래의 조건에 맞는 정보를 구하여 2-1의 DataFrame(df_target)에 입력하세요.
<br></br>
    - df_target의 '여자' Column Data의 Null값(None)을 채우기 위해 우선 위에서 구한 df_result값을 이용하려고 합니다.
<br></br>
    - 2-2의 df_result를 보면,
        - '회귀계수'의 경우 대부분 양수지만 아주 작은 값을 가지고 있습니다. 강동구의 경우 1년이 지날때 '여자비율'은 약 0.05% 정도 상승한다는 의미입니다.
        - 'MinMax'의 경우 최대 '여자비율'과 최소 '여자비율'의 차가 가장 큰 부분도 1.2%p 미만 수준입니다.
        - '표준편차'의 경우 대부분 평균의 0.5% 미만 수준으로 평균에 대부분의 '여자비율' Data들이 몰려있다고 보여집니다.
<br></br>
    - 따라서 '기간'(연도) 별로 '여자비율'은 평균에서 큰 차이가 없다고 판단되어 df_target의 '여자비율' Null값(None)에 df_result '구'('구분2') Data의 '평균' Data를 입력하려고 합니다.
<br></br>
    - 예시(df_result에서 '회귀계수'를 제외한 값은 실제와 값이 다름)
<br></br>
        <img src="images/06-2-3-001.png" style="margin:auto" width=800/>
<br></br>
<br></br>
    - 조건1: df_result의 '구'별 '평균' Data로 df_target의 '구분2'별 '여자비율' Data 중 Null값(None)을 대체하세요(위 예시 참고).
        - 소숫점 일곱번째 자리에서 반올림하여 여섯번째 자리까지만 나타내세요.
<br></br>
    - 조건2: 결과 DataFrame(df_target)의 Index, Column 순서 또는 Data 순서(정렬)는 변경하지 마세요.
<br></br>
    - 조건3: 'df_target' 변수에 결과 DataFrame을 할당하세요.
<br></br>
<br></br>
- 완료 후 결과 dataframe 변수(df_target)를 check_02_03 함수에 입력하여 채점하세요.

In [37]:
# df_result 재확인

df_result

구      회귀계수    MinMax         평균      표준편차
0    강동구  0.052052  0.701401  50.631950  0.222337
1    구로구  0.085093  1.011929  50.267907  0.293120
2    노원구  0.016932  0.400031  52.033941  0.153785
3    도봉구  0.024619  0.580787  51.265114  0.148100
4   동대문구  0.042682  0.488017  50.060631  0.141627
5    동작구  0.054593  0.531461  51.427075  0.186288
6   서대문구  0.059956  0.935272  51.826214  0.264601
7    서초구  0.066071  0.841456  52.552076  0.250892
8    성동구  0.097432  1.129327  50.518469  0.344289
9    송파구 -0.019510  0.285715  51.585111  0.095467
10   양천구  0.024970  0.378362  51.171848  0.124261
11  영등포구  0.082978  0.889680  50.260807  0.280208

In [38]:
# df_target의 여자비율 null 행 확인

df_target[(df_target.여자비율.isnull())]

기간 구분1   구분2    전체    남자  여자  여자비율
19   2009   구  동대문구  26.9  48.8 NaN   NaN
24   2009   구   노원구  22.3  43.4 NaN   NaN
33   2009   구   동작구  21.5  42.1 NaN   NaN
76   2010   구   송파구  20.7  42.0 NaN   NaN
111  2011   구   동작구  20.6  40.0 NaN   NaN
180  2013   구   노원구  20.8  40.7 NaN   NaN
184  2013   구   양천구  21.0  41.5 NaN   NaN
189  2013   구   동작구  20.8  40.7 NaN   NaN
223  2014   구   양천구  18.2  36.0 NaN   NaN
230  2014   구   서초구  14.7  29.0 NaN   NaN
251  2015   구   성동구  18.8  35.8 NaN   NaN
257  2015   구   도봉구  16.3  31.7 NaN   NaN
260  2015   구  서대문구  21.1  40.1 NaN   NaN
262  2015   구   양천구  17.0  33.0 NaN   NaN
306  2016   구   동작구  18.3  34.6 NaN   NaN
310  2016   구   송파구  16.9  32.8 NaN   NaN
342  2017   구   구로구  17.2  33.8 NaN   NaN
344  2017   구  영등포구  16.3  30.9 NaN   NaN
345  2017   구   동작구  16.8  32.2 NaN   NaN
347  2017   구   서초구  14.5  28.3 NaN   NaN
349  2017   구   송파구  17.2  34.4 NaN   NaN
383  2018   구  영등포구  15.7  29.5 NaN   NaN
388  2018   구   송파구  14.9  30.1 NaN   NaN
389  2018   구   강동구  15.6  29.8 NaN   NaN
427  2019   구   송파구  14.9  28.8 NaN   NaN

In [39]:
# if 문을 위한 check

pd.isnull(df_target.여자비율[1])

False

In [40]:
# if 문에 넣을 값 확인

df_result[df_result.구 == '동대문구']['평균']

4    50.060631
Name: 평균, dtype: float64

In [41]:
# 위 '구'에 넣을 것 확인

df_target.구분2[19]

'동대문구'

In [42]:
# for 문과 if 문 이용하여 nan 값 대체

for idx, rows in df_target.iterrows():
    if pd.isnull(df_target.여자비율[idx]):
        df_target.여자비율[idx] = df_result[df_result.구 == df_target['구분2'][idx]]['평균']
    else:
        continue

C:\Users\User\AppData\Local\Temp\ipykernel_18880\1979796579.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_target.여자비율[idx] = df_result[df_result.구 == df_target['구분2'][idx]]['평균']


In [43]:
# df_target의 여자비율 null 행 확인

df_target[(df_target.여자비율.isnull())]

Empty DataFrame
Columns: [기간, 구분1, 구분2, 전체, 남자, 여자, 여자비율]
Index: []

In [44]:
# null 값이었던 행 확인

df_target.loc[19]

기간           2009
구분1             구
구분2          동대문구
전체           26.9
남자           48.8
여자            NaN
여자비율    50.060631
Name: 19, dtype: object

In [45]:
# 2-3


check_02_03(df_target)
df_target.head()

정답입니다! 10점 누적 되었습니다!
현재 누적 점수: 50 / 100


기간    구분1     구분2    전체    남자   여자       여자비율
0  2009      시     서울시  24.3  45.5  4.0  51.084337
1  2009  생애주기별  19~29세  24.9  43.8  6.0  50.000000
2  2009  생애주기별  30~44세  29.9  55.1  3.8  49.122807
3  2009  생애주기별  45~64세  22.6  43.2  3.3  51.629073
4  2009  생애주기별  65세 이상  11.5  22.8  2.9  56.783920

### 문제 2-4) 심화 전처리 04 (10점) 
- 2-3의 DataFrame(df_target)을 아래의 조건에 맞게 변경하세요.
<br></br>
    - 이제까지 위에서 구했던 '전체', '남자', '여자비율' Column(열) Data를 활용하여 df_target의 '여자' Column Data의 Null값(None)을 채우려고 합니다.
<br></br>
    - **# df_target.전체 = df_target.여자비율/100 * x + (1 - df_target.여자비율/100) * df_target.남자** 에서의 x% 값이 '여자' 입니다.
<br></br>
<br></br>
    - 조건1: '여자' Column(열)의 Data가 Null값(None)인 경우 위의 수식에 따라 '여자'를 구하여 '여자' Column에 입력하세요.
        - '여자' Column(열)의 Data는 소숫점 두 번째 자리에서 반올림하여 소숫점 첫 번째까지 입력해주세요.
        - '여자' Column(열) Data의 Type은 float(numpy.float64)이어야 합니다.
<br></br>
    - 조건2: 조건1을 수행한 후 '여자비율' Column(열)을 삭제하세요.
<br></br>
    - 조건3: Index, Column 순서 또는 Data 순서(정렬)는 변경하지 마세요.
<br></br>
    - 조건4: 'df_target' 변수에 결과 DataFrame을 할당하세요.
<br></br>
<br></br>
    - hint1: 2-1과 같이 방정식을 풀기 위한 Python Library로 Sympy 등이 있습니다. 해당 문제는 sympy를 이용해서 풀수도 있고, 비교적 쉬운 1차 방정식이므로 수식을 직접 전개해서 풀어도 상관 없습니다.
<br></br>
<br></br>
- 완료 후 결과 dataframe 변수(df_target)를 check_02_04 함수에 입력하여 채점하세요.

In [46]:
# sympy test

x = sy.symbols('x')
f = sy.Eq(df_target.여자비율[0]/100 * x + (1 - df_target.여자비율[0]/100) * df_target.남자[0], df_target.전체[0])
result = sy.solve(f)
result

[3.99999971615566]

In [47]:
# for문에 sympy 적용하여 각 행에 결과 대입

x = sy.symbols('x')

for idx, rows in df_target.iterrows():
    if pd.isnull(df_target.여자[idx]):
        try:
            f = sy.Eq(df_target.여자비율[idx]/100 * x + (1 - df_target.여자비율[idx]/100) * df_target.남자[idx], df_target.전체[idx])
            df_target['여자'][idx] = sy.solve(f)[0]
        except:
            df_target['여자'][idx] = None


C:\Users\User\AppData\Local\Temp\ipykernel_18880\88063797.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_target['여자'][idx] = sy.solve(f)[0]


In [48]:
# null 행 확인

df_target[df_target.여자.isnull()]

Empty DataFrame
Columns: [기간, 구분1, 구분2, 전체, 남자, 여자, 여자비율]
Index: []

In [49]:
# 데이터 타입 변환 및 여자비율 드랍

df_target.여자 = df_target.여자.astype('float64').round(1)
df_target.drop(['여자비율'], axis=1, inplace=True)
df_target.head()

기간    구분1     구분2    전체    남자   여자
0  2009      시     서울시  24.3  45.5  4.0
1  2009  생애주기별  19~29세  24.9  43.8  6.0
2  2009  생애주기별  30~44세  29.9  55.1  3.8
3  2009  생애주기별  45~64세  22.6  43.2  3.3
4  2009  생애주기별  65세 이상  11.5  22.8  2.9

In [50]:
# 2-4


check_02_04(df_target)
df_target.head()

정답입니다! 10점 누적 되었습니다!
현재 누적 점수: 60 / 100


기간    구분1     구분2    전체    남자   여자
0  2009      시     서울시  24.3  45.5  4.0
1  2009  생애주기별  19~29세  24.9  43.8  6.0
2  2009  생애주기별  30~44세  29.9  55.1  3.8
3  2009  생애주기별  45~64세  22.6  43.2  3.3
4  2009  생애주기별  65세 이상  11.5  22.8  2.9

## 3단계: 원하는 정보 얻기
***

### 문제 3-1) 원하는 정보 얻기 01 (10점)
- 2단계의 DataFrame(df_target)을 이용하여 아래의 조건에 맞는 정보를 구하세요.
<br></br>
    - df_target '구분2' Column(열)의 '전체', '남자', '여자' 흡연율이 해마다 단조적(monotonic)으로 증가 또는 감소하는지 확인하고자 합니다.
<br></br>
    - 단조 증가 / 단조 감소
        - 어떤 수열이나 함수가 있을 때, 해당 수열이나 함수가 정의된 구간에서 감소하지 않는 경우를 단조증가, 증가하지 않는 경우를 단조감소 한다고 표현한다. 
        - 다시말해, 중간에 감소 한번 없이 지속해서 증가하는 경우 단조 증가, 중간에 증가 없이 지속해서 감소하는 경우 단조 감소한다고 한다.
<br></br>
<br></br>
    - 조건1: 결과 DataFrame의 Column(열)명은 ['구분1', '구분2', '전체', '남자','여자'] 입니다.
<br></br>
    - 조건2: '구분2' 별로 '전체', '남자', '여자' Column(열) Data가 '기간'에 따라 단조적으로 변화하는지 아닌지 확인하세요.
        - 단조적으로 변화한다는 것은 단조 증가 변화와 단조 감소 변화를 모두 포괄하는 말입니다.
        - 단조적으로 변화한다면 True, 그렇지 않다면 False를 입력하세요.
<br></br>
    - 조건3: ['구분1', '구분2']를 기준으로 오름차순으로 순서를 변경(sort)하고 Index를 Reset하세요.
<br></br>
    - 예시(임의의 boolean값을 넣었습니다.)
<br></br>
        <img src="images/06-3-1-001.png" style="margin:auto" width=400/> 
<br></br>
    - 조건4: 'df_result' 변수에 결과 DataFrame을 할당하세요.
<br></br>
<br></br>
- 완료 후 결과 dataframe 변수(df_result)를 check_03_01 함수에 입력하여 채점하세요.

In [51]:
# groupby, monotonic test
# monotonic >> https://hello70825.tistory.com/106

df_test = df_target.groupby(['구분1', '구분2']).agg(lambda x: x.is_monotonic)
df_test

C:\Users\User\AppData\Local\Temp\ipykernel_18880\1865407526.py:4: FutureWarning: is_monotonic is deprecated and will be removed in a future version. Use is_monotonic_increasing instead.
  df_test = df_target.groupby(['구분1', '구분2']).agg(lambda x: x.is_monotonic)


기간     전체     남자     여자
구분1            구분2                               
교육수준별Ⅰ(30~64세) 고졸       True  False  False  False
               대졸이상     True  False  False  False
               중졸이하     True  False  False  False
구              강남구      True  False  False  False
               강동구      True  False  False  False
               강북구      True  False  False  False
               강서구      True  False  False  False
               관악구      True  False  False  False
               광진구      True  False  False  False
               구로구      True  False  False  False
               금천구      True  False  False  False
               노원구      True  False  False  False
               도봉구      True  False  False  False
               동대문구     True  False  False  False
               동작구      True  False  False  False
               마포구      True  False  False  False
               서대문구     True  False  False  False
               서초구      True  False  False  False
               성동구      True  False  False  False
               성북구      True  False  False  False
               송파구      True  False  False  False
               양천구      True  False  False  False
               영등포구     True  False  False  False
               용산구      True  False  False  False
               은평구      True  False  False  False
               종로구      True  False  False  False
               중구       True  False  False  False
               중랑구      True  False  False  False
생애주기별          19~29세   True  False  False  False
               30~44세   True  False  False  False
               45~64세   True  False  False  False
               65세 이상   True  False  False  False
시              서울시      True  False  False  False
직업별(30~64세)    사무직      True  False  False  False
               서비스ㆍ판매직  True  False  False  False
               육체직      True  False  False  False

In [52]:
# '기간' 열 날리기

df_test = df_test[['전체', '남자', '여자']]
df_test

전체     남자     여자
구분1            구분2                         
교육수준별Ⅰ(30~64세) 고졸       False  False  False
               대졸이상     False  False  False
               중졸이하     False  False  False
구              강남구      False  False  False
               강동구      False  False  False
               강북구      False  False  False
               강서구      False  False  False
               관악구      False  False  False
               광진구      False  False  False
               구로구      False  False  False
               금천구      False  False  False
               노원구      False  False  False
               도봉구      False  False  False
               동대문구     False  False  False
               동작구      False  False  False
               마포구      False  False  False
               서대문구     False  False  False
               서초구      False  False  False
               성동구      False  False  False
               성북구      False  False  False
               송파구      False  False  False
               양천구      False  False  False
               영등포구     False  False  False
               용산구      False  False  False
               은평구      False  False  False
               종로구      False  False  False
               중구       False  False  False
               중랑구      False  False  False
생애주기별          19~29세   False  False  False
               30~44세   False  False  False
               45~64세   False  False  False
               65세 이상   False  False  False
시              서울시      False  False  False
직업별(30~64세)    사무직      False  False  False
               서비스ㆍ판매직  False  False  False
               육체직      False  False  False

In [53]:
# 인덱스 재정렬

df_test = df_test.reset_index()
df_test

구분1      구분2     전체     남자     여자
0   교육수준별Ⅰ(30~64세)       고졸  False  False  False
1   교육수준별Ⅰ(30~64세)     대졸이상  False  False  False
2   교육수준별Ⅰ(30~64세)     중졸이하  False  False  False
3                구      강남구  False  False  False
4                구      강동구  False  False  False
5                구      강북구  False  False  False
6                구      강서구  False  False  False
7                구      관악구  False  False  False
8                구      광진구  False  False  False
9                구      구로구  False  False  False
10               구      금천구  False  False  False
11               구      노원구  False  False  False
12               구      도봉구  False  False  False
13               구     동대문구  False  False  False
14               구      동작구  False  False  False
15               구      마포구  False  False  False
16               구     서대문구  False  False  False
17               구      서초구  False  False  False
18               구      성동구  False  False  False
19               구      성북구  False  False  False
20               구      송파구  False  False  False
21               구      양천구  False  False  False
22               구     영등포구  False  False  False
23               구      용산구  False  False  False
24               구      은평구  False  False  False
25               구      종로구  False  False  False
26               구       중구  False  False  False
27               구      중랑구  False  False  False
28           생애주기별   19~29세  False  False  False
29           생애주기별   30~44세  False  False  False
30           생애주기별   45~64세  False  False  False
31           생애주기별   65세 이상  False  False  False
32               시      서울시  False  False  False
33     직업별(30~64세)      사무직  False  False  False
34     직업별(30~64세)  서비스ㆍ판매직  False  False  False
35     직업별(30~64세)      육체직  False  False  False

In [54]:
# sort_values by 구분1, 구분2 오름차순

df_test.sort_values(by=['구분1', '구분2'], ascending=[True, True], inplace=True)
df_test

구분1      구분2     전체     남자     여자
0   교육수준별Ⅰ(30~64세)       고졸  False  False  False
1   교육수준별Ⅰ(30~64세)     대졸이상  False  False  False
2   교육수준별Ⅰ(30~64세)     중졸이하  False  False  False
3                구      강남구  False  False  False
4                구      강동구  False  False  False
5                구      강북구  False  False  False
6                구      강서구  False  False  False
7                구      관악구  False  False  False
8                구      광진구  False  False  False
9                구      구로구  False  False  False
10               구      금천구  False  False  False
11               구      노원구  False  False  False
12               구      도봉구  False  False  False
13               구     동대문구  False  False  False
14               구      동작구  False  False  False
15               구      마포구  False  False  False
16               구     서대문구  False  False  False
17               구      서초구  False  False  False
18               구      성동구  False  False  False
19               구      성북구  False  False  False
20               구      송파구  False  False  False
21               구      양천구  False  False  False
22               구     영등포구  False  False  False
23               구      용산구  False  False  False
24               구      은평구  False  False  False
25               구      종로구  False  False  False
26               구       중구  False  False  False
27               구      중랑구  False  False  False
28           생애주기별   19~29세  False  False  False
29           생애주기별   30~44세  False  False  False
30           생애주기별   45~64세  False  False  False
31           생애주기별   65세 이상  False  False  False
32               시      서울시  False  False  False
33     직업별(30~64세)      사무직  False  False  False
34     직업별(30~64세)  서비스ㆍ판매직  False  False  False
35     직업별(30~64세)      육체직  False  False  False

In [55]:
# 정렬까지 끝난 데이터프레임 df_result에 할당

df_result = df_test.copy()

In [56]:
# 3-1

check_03_01(df_result)
display(df_result)

정답입니다! 10점 누적 되었습니다!
현재 누적 점수: 70 / 100


구분1      구분2     전체     남자     여자
0   교육수준별Ⅰ(30~64세)       고졸  False  False  False
1   교육수준별Ⅰ(30~64세)     대졸이상  False  False  False
2   교육수준별Ⅰ(30~64세)     중졸이하  False  False  False
3                구      강남구  False  False  False
4                구      강동구  False  False  False
5                구      강북구  False  False  False
6                구      강서구  False  False  False
7                구      관악구  False  False  False
8                구      광진구  False  False  False
9                구      구로구  False  False  False
10               구      금천구  False  False  False
11               구      노원구  False  False  False
12               구      도봉구  False  False  False
13               구     동대문구  False  False  False
14               구      동작구  False  False  False
15               구      마포구  False  False  False
16               구     서대문구  False  False  False
17               구      서초구  False  False  False
18               구      성동구  False  False  False
19               구      성북구  False  False  False
20               구      송파구  False  False  False
21               구      양천구  False  False  False
22               구     영등포구  False  False  False
23               구      용산구  False  False  False
24               구      은평구  False  False  False
25               구      종로구  False  False  False
26               구       중구  False  False  False
27               구      중랑구  False  False  False
28           생애주기별   19~29세  False  False  False
29           생애주기별   30~44세  False  False  False
30           생애주기별   45~64세  False  False  False
31           생애주기별   65세 이상  False  False  False
32               시      서울시  False  False  False
33     직업별(30~64세)      사무직  False  False  False
34     직업별(30~64세)  서비스ㆍ판매직  False  False  False
35     직업별(30~64세)      육체직  False  False  False

### 문제 3-2) 원하는 정보 얻기 02 (10점)
- 2단계의 DataFrame(df_target)을 이용하여 아래의 조건에 맞는 정보를 구하세요.
<br></br>
    - df_target '구분2' Column(열)의 '전체', '남자', '여자' 흡연율이 해를 거듭할수록 증가 또는 감소하는지 회귀계수를 통해서 확인하고자 합니다.
<br></br>
    - 회귀계수에 대해서는 앞의 2-2 문제를 참고해주세요.
<br></br>
<br></br>
    - 조건1: 결과 DataFrame의 Column(열)명은 ['구분1', '구분2', '전체', '남자','여자'] 입니다.
<br></br>
    - 조건2: '구분2' 별로 '전체', '남자', '여자' Column(열) Data에 대해 '기간'에 따른 회귀계수를 구하세요.
<br></br>
    - 조건3: ['구분1', '구분2']를 기준으로 오름차순으로 순서를 변경(sort)하고 Index를 Reset하세요.
<br></br>
    - 예시(임의의 float값을 넣었습니다.)
<br></br>
        <img src="images/06-3-2-001.png" style="margin:auto" width=400/> 
<br></br>
    - 조건4: 'df_result' 변수에 결과 DataFrame을 할당하세요.
<br></br>
<br></br>
- 완료 후 결과 dataframe 변수(df_result)를 check_03_02 함수에 입력하여 채점하세요.

In [57]:
# 구분1, 2, 기간으로 그룹화 후 인덱스 재생성
# 이후 구분2 별로 tmp dataframe 이용하여 reg 구할 예정

df_result = df_target.groupby(['구분1', '구분2', '기간']).mean()
df_result = df_result.reset_index()
df_result

구분1  구분2    기간    전체    남자   여자
0    교육수준별Ⅰ(30~64세)   고졸  2009  27.7  54.5  4.4
1    교육수준별Ⅰ(30~64세)   고졸  2010  27.0  53.8  4.7
2    교육수준별Ⅰ(30~64세)   고졸  2011  27.7  54.4  4.6
3    교육수준별Ⅰ(30~64세)   고졸  2012  27.1  54.1  4.4
4    교육수준별Ⅰ(30~64세)   고졸  2013  27.4  55.3  4.9
..              ...  ...   ...   ...   ...  ...
391     직업별(30~64세)  육체직  2015  34.8  48.0  3.9
392     직업별(30~64세)  육체직  2016  34.2  47.7  3.5
393     직업별(30~64세)  육체직  2017  34.7  49.0  2.5
394     직업별(30~64세)  육체직  2018  33.2  46.5  5.4
395     직업별(30~64세)  육체직  2019  34.2  46.5  4.3

[396 rows x 6 columns]

In [58]:
# 구분2 리스트 생성

category = list(df_result.구분2.unique())
category

['고졸',
 '대졸이상',
 '중졸이하',
 '강남구',
 '강동구',
 '강북구',
 '강서구',
 '관악구',
 '광진구',
 '구로구',
 '금천구',
 '노원구',
 '도봉구',
 '동대문구',
 '동작구',
 '마포구',
 '서대문구',
 '서초구',
 '성동구',
 '성북구',
 '송파구',
 '양천구',
 '영등포구',
 '용산구',
 '은평구',
 '종로구',
 '중구',
 '중랑구',
 '19~29세',
 '30~44세',
 '45~64세',
 '65세 이상',
 '서울시',
 '사무직',
 '서비스ㆍ판매직',
 '육체직']

In [59]:
# test

df_result[df_result.구분2 == category[0]]

구분1 구분2    기간    전체    남자   여자
0   교육수준별Ⅰ(30~64세)  고졸  2009  27.7  54.5  4.4
1   교육수준별Ⅰ(30~64세)  고졸  2010  27.0  53.8  4.7
2   교육수준별Ⅰ(30~64세)  고졸  2011  27.7  54.4  4.6
3   교육수준별Ⅰ(30~64세)  고졸  2012  27.1  54.1  4.4
4   교육수준별Ⅰ(30~64세)  고졸  2013  27.4  55.3  4.9
5   교육수준별Ⅰ(30~64세)  고졸  2014  26.8  52.8  4.9
6   교육수준별Ⅰ(30~64세)  고졸  2015  26.1  50.5  5.2
7   교육수준별Ⅰ(30~64세)  고졸  2016  25.7  51.3  4.8
8   교육수준별Ⅰ(30~64세)  고졸  2017  25.3  49.1  5.8
9   교육수준별Ⅰ(30~64세)  고졸  2018  24.1  48.2  4.7
10  교육수준별Ⅰ(30~64세)  고졸  2019  24.2  46.7  5.1

In [60]:
# for문 이용하여 진행

reg_tot = []
reg_male = []
reg_female = []

for i in category:
    df_tmp = df_result[df_result.구분2 == i]

    fp1 = np.polyfit(df_tmp.기간, df_tmp.전체, 1)
    fp2 = np.polyfit(df_tmp.기간, df_tmp.남자, 1)
    fp3 = np.polyfit(df_tmp.기간, df_tmp.여자, 1)

    reg_tot.append(fp1[0])
    reg_male.append(fp2[0])
    reg_female.append(fp3[0])
    

In [61]:
# 리스트 확인

reg_tot[:20], len(reg_tot)

([-0.36727272727270527,
  -0.8727272727272251,
  0.006363636363617508,
  -0.4881818181817869,
  -0.9090909090908538,
  -0.5227272727272445,
  -0.5072727272727136,
  -0.5836363636363493,
  -0.4854545454545116,
  -0.7945454545454039,
  -0.6354545454545073,
  -0.5299999999999653,
  -0.7627272727272175,
  -0.8545454545454054,
  -0.6490909090908642,
  -0.9072727272726913,
  -0.5427272727272486,
  -0.5872727272727024,
  -0.5563636363636041,
  -0.8345454545453973],
 36)

In [62]:
# 데이터프레임 groupby / reset_index로 자료 넣기 쉽게 변형
# 기간 열 제거

df_result = df_result.groupby(['구분1', '구분2']).mean().reset_index()
df_result = df_result.drop('기간', axis =1)
df_result

구분1      구분2         전체         남자        여자
0   교육수준별Ⅰ(30~64세)       고졸  26.281818  51.881818  4.863636
1   교육수준별Ⅰ(30~64세)     대졸이상  21.781818  38.436364  2.027273
2   교육수준별Ⅰ(30~64세)     중졸이하  20.618182  46.600000  4.472727
3                구      강남구  18.281818  33.772727  4.600000
4                구      강동구  20.618182  37.863636  3.800000
5                구      강북구  23.736364  43.563636  4.763636
6                구      강서구  20.727273  39.372727  3.227273
7                구      관악구  22.581818  40.418182  4.427273
8                구      광진구  21.190909  39.254545  3.972727
9                구      구로구  20.727273  38.763636  2.863636
10               구      금천구  23.663636  43.009091  3.581818
11               구      노원구  19.963636  38.354545  3.009091
12               구      도봉구  21.318182  39.709091  3.827273
13               구     동대문구  22.163636  40.272727  4.090909
14               구      동작구  19.327273  36.672727  2.936364
15               구      마포구  20.681818  38.363636  4.572727
16               구     서대문구  20.981818  39.754545  3.518182
17               구      서초구  16.681818  32.463636  2.418182
18               구      성동구  21.681818  40.072727  3.645455
19               구      성북구  22.272727  42.018182  3.563636
20               구      송파구  18.663636  36.100000  2.309091
21               구      양천구  19.300000  37.236364  2.172727
22               구     영등포구  19.581818  36.654545  2.672727
23               구      용산구  20.972727  38.981818  4.336364
24               구      은평구  21.390909  39.345455  4.563636
25               구      종로구  21.172727  38.790909  4.154545
26               구       중구  23.081818  41.845455  4.481818
27               구      중랑구  22.709091  41.672727  3.972727
28           생애주기별   19~29세  20.854545  36.263636  5.463636
29           생애주기별   30~44세  25.890909  47.318182  3.909091
30           생애주기별   45~64세  20.363636  39.236364  2.954545
31           생애주기별   65세 이상   9.900000  19.381818  2.163636
32               시      서울시  20.700000  38.681818  3.572727
33     직업별(30~64세)      사무직  24.990909  38.318182  2.718182
34     직업별(30~64세)  서비스ㆍ판매직  28.100000  49.918182  6.327273
35     직업별(30~64세)      육체직  35.809091  49.918182  4.009091

In [63]:
# 위에서 만든 list 형태의 regression 데이터 기입

df_result.전체 = reg_tot
df_result.남자 = reg_male
df_result.여자 = reg_female

df_result

구분1      구분2        전체        남자        여자
0   교육수준별Ⅰ(30~64세)       고졸 -0.367273 -0.797273  0.074545
1   교육수준별Ⅰ(30~64세)     대졸이상 -0.872727 -1.285455 -0.013636
2   교육수준별Ⅰ(30~64세)     중졸이하  0.006364 -0.205455  0.075455
3                구      강남구 -0.488182 -0.910909 -0.124545
4                구      강동구 -0.909091 -1.543636 -0.260000
5                구      강북구 -0.522727 -0.956364 -0.062727
6                구      강서구 -0.507273 -1.085455  0.082727
7                구      관악구 -0.583636 -1.073636 -0.057273
8                구      광진구 -0.485455 -1.033636  0.107273
9                구      구로구 -0.794545 -1.485455 -0.056364
10               구      금천구 -0.635455 -1.273636  0.038182
11               구      노원구 -0.530000 -1.214545  0.110000
12               구      도봉구 -0.762727 -1.647273  0.091818
13               구     동대문구 -0.854545 -1.599091 -0.092727
14               구      동작구 -0.649091 -1.423636  0.101818
15               구      마포구 -0.907273 -1.665455 -0.141818
16               구     서대문구 -0.542727 -1.038182 -0.043636
17               구      서초구 -0.587273 -1.049091 -0.137273
18               구      성동구 -0.556364 -1.060000  0.004545
19               구      성북구 -0.834545 -1.621818 -0.033636
20               구      송파구 -0.782727 -1.517273 -0.093636
21               구      양천구 -0.705455 -1.420909 -0.007273
22               구     영등포구 -0.783636 -1.387273 -0.147273
23               구      용산구 -0.831818 -1.455455 -0.222727
24               구      은평구 -0.619091 -1.361818  0.137273
25               구      종로구 -0.641818 -1.177273 -0.073636
26               구       중구 -0.701818 -1.181818 -0.150909
27               구      중랑구 -0.333636 -0.734545  0.086364
28           생애주기별   19~29세 -0.657273 -1.416364  0.141818
29           생애주기별   30~44세 -0.940000 -1.802727 -0.008182
30           생애주기별   45~64세 -0.252727 -0.465455 -0.031818
31           생애주기별   65세 이상 -0.310000 -0.560909 -0.153636
32               시      서울시 -0.660000 -1.286364 -0.027273
33     직업별(30~64세)      사무직 -1.218182 -1.408182 -0.053636
34     직업별(30~64세)  서비스ㆍ판매직 -0.701818 -0.876364  0.012727
35     직업별(30~64세)      육체직 -0.528182 -0.790909 -0.009091

In [64]:
# 소수점 7번째자리 반올림

df_result.전체 = df_result.전체.round(6)
df_result.남자 = df_result.남자.round(6)
df_result.여자 = df_result.여자.round(6)
df_result

구분1      구분2        전체        남자        여자
0   교육수준별Ⅰ(30~64세)       고졸 -0.367273 -0.797273  0.074545
1   교육수준별Ⅰ(30~64세)     대졸이상 -0.872727 -1.285455 -0.013636
2   교육수준별Ⅰ(30~64세)     중졸이하  0.006364 -0.205455  0.075455
3                구      강남구 -0.488182 -0.910909 -0.124545
4                구      강동구 -0.909091 -1.543636 -0.260000
5                구      강북구 -0.522727 -0.956364 -0.062727
6                구      강서구 -0.507273 -1.085455  0.082727
7                구      관악구 -0.583636 -1.073636 -0.057273
8                구      광진구 -0.485455 -1.033636  0.107273
9                구      구로구 -0.794545 -1.485455 -0.056364
10               구      금천구 -0.635455 -1.273636  0.038182
11               구      노원구 -0.530000 -1.214545  0.110000
12               구      도봉구 -0.762727 -1.647273  0.091818
13               구     동대문구 -0.854545 -1.599091 -0.092727
14               구      동작구 -0.649091 -1.423636  0.101818
15               구      마포구 -0.907273 -1.665455 -0.141818
16               구     서대문구 -0.542727 -1.038182 -0.043636
17               구      서초구 -0.587273 -1.049091 -0.137273
18               구      성동구 -0.556364 -1.060000  0.004545
19               구      성북구 -0.834545 -1.621818 -0.033636
20               구      송파구 -0.782727 -1.517273 -0.093636
21               구      양천구 -0.705455 -1.420909 -0.007273
22               구     영등포구 -0.783636 -1.387273 -0.147273
23               구      용산구 -0.831818 -1.455455 -0.222727
24               구      은평구 -0.619091 -1.361818  0.137273
25               구      종로구 -0.641818 -1.177273 -0.073636
26               구       중구 -0.701818 -1.181818 -0.150909
27               구      중랑구 -0.333636 -0.734545  0.086364
28           생애주기별   19~29세 -0.657273 -1.416364  0.141818
29           생애주기별   30~44세 -0.940000 -1.802727 -0.008182
30           생애주기별   45~64세 -0.252727 -0.465455 -0.031818
31           생애주기별   65세 이상 -0.310000 -0.560909 -0.153636
32               시      서울시 -0.660000 -1.286364 -0.027273
33     직업별(30~64세)      사무직 -1.218182 -1.408182 -0.053636
34     직업별(30~64세)  서비스ㆍ판매직 -0.701818 -0.876364  0.012727
35     직업별(30~64세)      육체직 -0.528182 -0.790909 -0.009091

In [65]:
# 3-2

check_03_02(df_result)

정답입니다! 10점 누적 되었습니다!
현재 누적 점수: 80 / 100


### 문제 3-3) 원하는 정보 얻기 03 (10점)
- 3-2의 DataFrame(df_result)을 이용하여 아래의 조건에 맞는 정보를 구하세요.
<br></br>
    - 3-2의 df_result '구분1' Column(열) Data마다 어떤 '구분2' Column(열)의 Data가 '전체' 흡연율 감소율이 높은지/낮은지 알아보고자 합니다.
<br></br>
<br></br>
    - 조건1: 결과 DataFrame의 Column(열)명은 ['구분1', '최대', '최소'] 입니다.
<br></br>
    - 조건2: 회귀계수가 음수(=흡연율이 감소중)인 '구분1' Column(열) Data 중 어떤 '구분2' Column(열) Data가 '전체' 흡연율 감소율이 가장 높은지/낮은지 찾아서 결과 DataFrame에 입력하세요.
        - '구분1' 중 '시'는 '구분2' Data 가 1개(서울시)이므로 제외합니다.
        - 만약, '구분1' Column(열) Data 중 감소중인 '구분2' Column(열) Data가 1개라면, 해당 '구분2' Data를 '최대'에 넣고, '최소' Column(열)에는 None값을 입력하세요.
        - 만약, '구분1' Column(열) Data 중 증가중인 '구분2' Column(열) Data가 없다면, 해당 '구분1' Data는 건너뛰세요.(결과 DataFrame에 입력하지 마세요)
<br></br>
    - 조건3: ['구분1']을 기준으로 오름차순으로 순서를 변경(sort)하고 Index를 Reset하세요.
<br></br>
    - 예시(임의의 값을 넣었습니다.)
<br></br>
        <img src="images/06-3-3-001.png" style="margin:auto" width=300/> 
<br></br>
    - 조건4: 'df_maxmin' 변수에 결과 DataFrame을 할당하세요.
<br></br>
<br></br>
- 완료 후 결과 dataframe 변수(df_maxmin)를 check_03_04 함수에 입력하여 채점하세요.

In [66]:
# 구분1 유니크로 데이터 리스트 생성
# 시 제거

cate1 = list(df_result.구분1.unique())
cate1.remove('시')
cate1

['교육수준별Ⅰ(30~64세)', '구', '생애주기별', '직업별(30~64세)']

In [67]:
# tmp test

df_tmp = df_result[df_result.구분1 == cate1[0]]
df_tmp

구분1   구분2        전체        남자        여자
0  교육수준별Ⅰ(30~64세)    고졸 -0.367273 -0.797273  0.074545
1  교육수준별Ⅰ(30~64세)  대졸이상 -0.872727 -1.285455 -0.013636
2  교육수준별Ⅰ(30~64세)  중졸이하  0.006364 -0.205455  0.075455

In [68]:
# max 값 확인

df_tmp.전체.max()

0.006364

In [69]:
# 위 max 값으로 이전 문제 데이터셋의 해당 행 인덱스 가져오기

df_result[df_result.전체==df_tmp.전체.max()].index[0]

2

In [70]:
# 위에서 가져온 인덱스를 이용하여 구분2 값 가져오기

df_result.loc[df_result[df_result.전체==df_tmp.전체.max()].index[0], '구분2']

'중졸이하'

In [71]:
# drop 테스트 (1)

df_tmp

구분1   구분2        전체        남자        여자
0  교육수준별Ⅰ(30~64세)    고졸 -0.367273 -0.797273  0.074545
1  교육수준별Ⅰ(30~64세)  대졸이상 -0.872727 -1.285455 -0.013636
2  교육수준별Ⅰ(30~64세)  중졸이하  0.006364 -0.205455  0.075455

In [72]:
# drop 테스트 (2)

df_tmp.drop(df_tmp[df_tmp.전체 == df_tmp.전체.max()].index[0])

구분1   구분2        전체        남자        여자
0  교육수준별Ⅰ(30~64세)    고졸 -0.367273 -0.797273  0.074545
1  교육수준별Ⅰ(30~64세)  대졸이상 -0.872727 -1.285455 -0.013636

In [73]:
# for문 이용하여 흡연율 감소 최대/최소 구하기
# 데이터 중 감소가 1개인 데이터는 없기에 여타 다른 조건은 skip
# 오히려 증가하고 있는 데이터는 if 문 사용하여 해당 행을 날리는 방법 이용

mini = []
maxi = []

for i in cate1:
    df_tmp = df_result[df_result.구분1 == i]
    if df_tmp.전체.max() > 0:
        df_tmp = df_tmp.drop(df_tmp[df_tmp.전체 == df_tmp.전체.max()].index[0]) # 흡연율 증가되는 행의 인덱스 찾아서 drop하는 방식
    maxi.append(df_result.loc[df_result[df_result.전체==df_tmp.전체.max()].index[0], '구분2']) # << min(), max() 반대로 되어야 하는 부분
    mini.append(df_result.loc[df_result[df_result.전체==df_tmp.전체.min()].index[0], '구분2']) # << min(), max() 반대로 되어야 하는 부분


In [74]:
# 위에서 추출한 최소, 최대 리스트 확인

mini, maxi

(['대졸이상', '강동구', '30~44세', '사무직'], ['고졸', '중랑구', '45~64세', '육체직'])

In [75]:
# 데이터프레임 생성

df_maxmin = pd.DataFrame({'구분1':cate1, '최대':maxi, '최소':mini})
df_maxmin

구분1      최대      최소
0  교육수준별Ⅰ(30~64세)      고졸    대졸이상
1               구     중랑구     강동구
2           생애주기별  45~64세  30~44세
3     직업별(30~64세)     육체직     사무직

In [76]:
# 3-3

check_03_03(df_maxmin)
df_maxmin

정답입니다! 10점 누적 되었습니다!
현재 누적 점수: 90 / 100


구분1      최대      최소
0  교육수준별Ⅰ(30~64세)      고졸    대졸이상
1               구     중랑구     강동구
2           생애주기별  45~64세  30~44세
3     직업별(30~64세)     육체직     사무직

### 문제 3-4) 원하는 정보 얻기 04 (10점)
- 3-2의 DataFrame(df_result)을 이용하여 아래의 조건에 맞는 정보를 구하세요.
<br></br>
    - 3-2의 df_result '구분1' Column(열) Data마다 어떤 '구분2' Column(열)의 Data가 '여자' 흡연율 증가율이 높은지/낮은지 알아보고자 합니다.
<br></br>
<br></br>
    - 조건1: 결과 DataFrame의 Column(열)명은 ['구분1', '최대', '최소'] 입니다.
<br></br>
    - 조건2: 회귀계수가 양수(=흡연율이 증가중)인 '구분1' Column(열) Data 중 어떤 '구분2' Column(열) Data가 '여자' 흡연율 감소율이 가장 높은지/낮은지 찾아서 결과 DataFrame에 입력하세요.
        - '구분1' 중 '시'는 '구분2' Data 가 1개(서울시)이므로 제외합니다.
        - 만약, '구분1' Column(열) Data 중 증가중인 '구분2' Column(열) Data가 1개라면, 해당 '구분2' Data를 '최대'에 넣고, '최소' Column(열)에는 None값을 입력하세요.
        - 만약, '구분1' Column(열) Data 중 증가중인 '구분2' Column(열) Data가 없다면, 해당 '구분1' Data는 건너뛰세요.(결과 DataFrame에 입력하지 마세요)
<br></br>
    - 조건3: ['구분1']을 기준으로 오름차순으로 순서를 변경(sort)하고 Index를 Reset하세요.
<br></br>
    - 예시(임의의 값을 넣었습니다.)
<br></br>
        <img src="images/06-3-4-001.png" style="margin:auto" width=300/> 
<br></br>
    - 조건4: 'df_maxmin' 변수에 결과 DataFrame을 할당하세요.
<br></br>
<br></br>
- 완료 후 결과 dataframe 변수(df_maxmin)를 check_03_04 함수에 입력하여 채점하세요.

In [77]:
# 원본 데이터 확인

df_result

구분1      구분2        전체        남자        여자
0   교육수준별Ⅰ(30~64세)       고졸 -0.367273 -0.797273  0.074545
1   교육수준별Ⅰ(30~64세)     대졸이상 -0.872727 -1.285455 -0.013636
2   교육수준별Ⅰ(30~64세)     중졸이하  0.006364 -0.205455  0.075455
3                구      강남구 -0.488182 -0.910909 -0.124545
4                구      강동구 -0.909091 -1.543636 -0.260000
5                구      강북구 -0.522727 -0.956364 -0.062727
6                구      강서구 -0.507273 -1.085455  0.082727
7                구      관악구 -0.583636 -1.073636 -0.057273
8                구      광진구 -0.485455 -1.033636  0.107273
9                구      구로구 -0.794545 -1.485455 -0.056364
10               구      금천구 -0.635455 -1.273636  0.038182
11               구      노원구 -0.530000 -1.214545  0.110000
12               구      도봉구 -0.762727 -1.647273  0.091818
13               구     동대문구 -0.854545 -1.599091 -0.092727
14               구      동작구 -0.649091 -1.423636  0.101818
15               구      마포구 -0.907273 -1.665455 -0.141818
16               구     서대문구 -0.542727 -1.038182 -0.043636
17               구      서초구 -0.587273 -1.049091 -0.137273
18               구      성동구 -0.556364 -1.060000  0.004545
19               구      성북구 -0.834545 -1.621818 -0.033636
20               구      송파구 -0.782727 -1.517273 -0.093636
21               구      양천구 -0.705455 -1.420909 -0.007273
22               구     영등포구 -0.783636 -1.387273 -0.147273
23               구      용산구 -0.831818 -1.455455 -0.222727
24               구      은평구 -0.619091 -1.361818  0.137273
25               구      종로구 -0.641818 -1.177273 -0.073636
26               구       중구 -0.701818 -1.181818 -0.150909
27               구      중랑구 -0.333636 -0.734545  0.086364
28           생애주기별   19~29세 -0.657273 -1.416364  0.141818
29           생애주기별   30~44세 -0.940000 -1.802727 -0.008182
30           생애주기별   45~64세 -0.252727 -0.465455 -0.031818
31           생애주기별   65세 이상 -0.310000 -0.560909 -0.153636
32               시      서울시 -0.660000 -1.286364 -0.027273
33     직업별(30~64세)      사무직 -1.218182 -1.408182 -0.053636
34     직업별(30~64세)  서비스ㆍ판매직 -0.701818 -0.876364  0.012727
35     직업별(30~64세)      육체직 -0.528182 -0.790909 -0.009091

In [78]:
# 양수 데이터만 고르기

df_tmp = df_tmp[df_tmp.여자 >0]
df_tmp

구분1      구분2        전체        남자        여자
34  직업별(30~64세)  서비스ㆍ판매직 -0.701818 -0.876364  0.012727

In [79]:
# for문 활용하여 list 생성

mini = []
maxi = []

for i in cate1:
    df_tmp = df_result[df_result.구분1 == i]
    df_tmp = df_tmp[df_tmp.여자 >0]
    if len(df_tmp) == 1:
        maxi.append(df_result.loc[df_result[df_result.여자==df_tmp.여자.max()].index[0], '구분2'])
        mini.append(None)
        continue        
    maxi.append(df_result.loc[df_result[df_result.여자==df_tmp.여자.max()].index[0], '구분2'])
    mini.append(df_result.loc[df_result[df_result.여자==df_tmp.여자.min()].index[0], '구분2'])


In [80]:
# 데이터 리스트 확인

mini, maxi

(['고졸', '성동구', None, None], ['중졸이하', '은평구', '19~29세', '서비스ㆍ판매직'])

In [81]:
df_maxmin = pd.DataFrame({'구분1':cate1, '최대':maxi, '최소':mini})
df_maxmin

구분1       최대    최소
0  교육수준별Ⅰ(30~64세)     중졸이하    고졸
1               구      은평구   성동구
2           생애주기별   19~29세  None
3     직업별(30~64세)  서비스ㆍ판매직  None

In [82]:
df_maxmin.iloc[[0]].index

Int64Index([0], dtype='int64')

In [83]:
df_maxmin = df_maxmin.drop(df_maxmin.iloc[[0]].index, axis=0)
df_maxmin

구분1       최대    최소
1            구      은평구   성동구
2        생애주기별   19~29세  None
3  직업별(30~64세)  서비스ㆍ판매직  None

In [84]:
df_maxmin.reset_index(drop=True, inplace=True)
df_maxmin

구분1       최대    최소
0            구      은평구   성동구
1        생애주기별   19~29세  None
2  직업별(30~64세)  서비스ㆍ판매직  None

In [85]:
# 3-4

check_03_04(df_maxmin)
df_maxmin

정답입니다! 10점 누적 되었습니다!
현재 누적 점수: 100 / 100


구분1       최대    최소
0            구      은평구   성동구
1        생애주기별   19~29세  None
2  직업별(30~64세)  서비스ㆍ판매직  None

***

### 🎊 문제가 끝났습니다. 제출양식과 일정에 맞춰 제출해주시기를 바랍니다!

### 제출양식
- 이 파일은 EDA TEST 6회차 노트북입니다. 따라서 제출시 파일명은 아래와 같습니다. 
- [DS]edalvtest_06_yourname.ipynb

***

# 문의사항 및 의견이 있는 경우 작성해주세요. 

#### 문제 수정 요청

* 3-2 소수점 7번째 자리에서 반올림하라는 조건이 없어 혼동 여지가 있습니다. 반올림 조건 추가 부탁드립니다.


#### 답안 수정 요청

* 3-3 답안의 최대 최소가 바뀐 것 같습니다.
  - 고졸의 흡연율은 -0.367273 / 대졸이상의 흡연율은	-0.872727로 대졸이상이 흡연율 감소율이 최대로 높습니다.
  - 답안에는 흡연율 감소율 최대가 고졸, 최소가 대졸로 되어있는데, min() / max() 를 반대로 넣어야 될 것으로 보입니다.

<br>

* 3-4 답안에도 문제가 있습니다.
  - 교육수준별Ⅰ(30~64세)	고졸	0.074545
  - 교육수준별Ⅰ(30~64세)	대졸이상	-0.013636
  - 교육수준별Ⅰ(30~64세)	중졸이하	0.075455
  - 위에서 흡연율이 증가하고 있는 행이 두개가 있는데, 해당 부분이 답안에 사라져 있습니다.

<br>

* 일단 답안에 맞춰 작성해야 될 것으로 생각되어 답안과 똑같이 나오도록 코드를 짰으나, 다음 수강생들을 위해 답안 수정 요청드립니다.